File: src/shared/detection.ts
import type { ClipboardFormatInfo, ClipboardItemType, ClipboardPlatform, CodeLanguage, ContentSignal } from './types';

const urlPattern = /^(https?:\/\/|www\.)[^\s/$.?#].[^\s]*$/i;
const embeddedUrlPattern = /\bhttps?:\/\/[^\s<>"']+|\bwww\.[^\s<>"']+/i;
const emailPattern = /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/i;
const hexColorPattern = /(^|[^A-Fa-f0-9])#(?:[A-Fa-f0-9]{3}|[A-Fa-f0-9]{6}|[A-Fa-f0-9]{8})(?![A-Fa-f0-9])/;
const shellPattern = /^\s*(npm|pnpm|yarn|git|cd|ls|mkdir|rm|cp|mv|curl|wget|docker|kubectl|ssh|scp|chmod|chown|brew|pip|python|node)\b/m;
const markdownPattern = /(^|\n)\s{0,3}(#{1,6}\s+\S|[-*+]\s+\S|\d+\.\s+\S|>\s+\S|```)/;

export function isUrl(text: string): boolean {
  return urlPattern.test(text.trim());
}

export function detectCodeLanguage(text: string): CodeLanguage {
  const trimmed = text.trim();
  if (!trimmed) return 'unknown';
  if ((trimmed.startsWith('{') || trimmed.startsWith('[')) && isJson(trimmed)) return 'json';
  if (/<!doctype html|<\/?[a-z][\s\S]*>/i.test(trimmed)) return 'html';
  if (/^\s*([.#]?[\w-]+\s*\{|@media|:root\s*\{)/m.test(trimmed) && /[{}:;]/.test(trimmed)) return 'css';
  if (/^\s*(def|import|from|class)\s+\w+|print\(|if __name__ == ['"]__main__['"]/m.test(trimmed)) return 'python';
  if (isCommandSnippet(trimmed)) return 'shell';
  if (/\b(interface|type|enum|implements|readonly)\b/.test(trimmed) && /\b(import|export|const|let|function)\b/.test(trimmed)) return 'typescript';
  if (/\b(import|export|const|let|function|=>|console\.log|require\()\b/.test(trimmed)) return 'javascript';
  return 'unknown';
}

export function isCommandSnippet(text: string): boolean {
  const lines = text.trim().split('\n').map((line) => line.trim()).filter(Boolean);
  return lines.length > 0 && lines.length <= 4 && lines.every((line) => shellPattern.test(line)) && !/[{};]/.test(text.replace(/;?\s*$/, ''));
}

export function detectClipboardType(input: {
  text?: string;
  html?: string;
  rtf?: string;
  hasImage?: boolean;
  filePaths?: string[];
}): { type: ClipboardItemType; codeLanguage: CodeLanguage | null; url: string | null } {
  if (input.hasImage) return { type: 'image', codeLanguage: null, url: null };
  if (input.filePaths?.length) return { type: 'file_reference', codeLanguage: null, url: null };

  const text = input.text?.trim() ?? '';
  if (text && isUrl(text)) return { type: 'url', codeLanguage: null, url: text.startsWith('www.') ? `https://${text}` : text };
  if (text && isCommandSnippet(text)) return { type: 'command', codeLanguage: 'shell', url: null };

  const language = text ? detectCodeLanguage(text) : 'unknown';
  if (language !== 'unknown') return { type: 'code', codeLanguage: language, url: null };
  if (input.html || input.rtf) return { type: 'rich_text', codeLanguage: null, url: null };
  return { type: 'plain_text', codeLanguage: null, url: null };
}

export function normalizeClipboardFormats(input: {
  rawFormats?: string[];
  hasText?: boolean;
  hasHtml?: boolean;
  hasRtf?: boolean;
  hasImage?: boolean;
  hasFiles?: boolean;
  platform?: NodeJS.Platform | ClipboardPlatform;
}): ClipboardFormatInfo {
  const rawFormats = Array.from(new Set((input.rawFormats ?? []).filter(Boolean)));
  const lowerFormats = rawFormats.map((format) => format.toLowerCase());
  const normalized = new Set<string>();

  if (input.hasText || lowerFormats.some((format) => /text\/plain|plain-text|utf8-plain-text|unicode text|cf_unicode/.test(format))) normalized.add('text/plain');
  if (input.hasHtml || lowerFormats.some((format) => /text\/html|public\.html|html format/.test(format))) normalized.add('text/html');
  if (input.hasRtf || lowerFormats.some((format) => /text\/rtf|public\.rtf|rich text|rtf/.test(format))) normalized.add('text/rtf');
  if (input.hasImage || lowerFormats.some((format) => /image\/|public\.(png|jpeg|tiff)|bitmap|dib|cf_dib/.test(format))) normalized.add('image');
  if (input.hasFiles || lowerFormats.some((format) => /file|uri-list|filename|nsfilenames|cf_hdrop/.test(format))) normalized.add('files');

  return {
    rawFormats,
    normalizedFormats: Array.from(normalized),
    hasText: normalized.has('text/plain'),
    hasHtml: normalized.has('text/html'),
    hasRtf: normalized.has('text/rtf'),
    hasImage: normalized.has('image'),
    hasFiles: normalized.has('files'),
    platform: normalizePlatform(input.platform),
  };
}

export function detectContentSignals(input: {
  text?: string;
  html?: string;
  rtf?: string;
  hasImage?: boolean;
  filePaths?: string[];
  codeLanguage?: CodeLanguage | null;
}): ContentSignal[] {
  const signals: ContentSignal[] = [];
  const text = input.text ?? '';
  const trimmed = text.trim();

  if (input.html?.trim()) signals.push({ kind: 'html', confidence: 'high' });
  if (input.codeLanguage && input.codeLanguage !== 'unknown') {
    signals.push({ kind: input.codeLanguage === 'shell' ? 'shell' : 'code', confidence: 'high', language: input.codeLanguage });
    if (input.codeLanguage === 'json') signals.push({ kind: 'json', confidence: 'high', language: 'json', range: { start: text.indexOf(trimmed), end: text.indexOf(trimmed) + trimmed.length } });
  }

  if (!trimmed) return dedupeSignals(signals);

  if (!signals.some((signal) => signal.kind === 'json')) {
    const jsonFragment = findJsonFragment(text);
    if (jsonFragment) signals.push({ kind: 'json_fragment', confidence: 'medium', language: 'json', range: jsonFragment });
  }

  const urlMatch = text.match(embeddedUrlPattern);
  if (urlMatch?.index !== undefined) signals.push({ kind: 'url', confidence: isUrl(trimmed) ? 'high' : 'medium', range: { start: urlMatch.index, end: urlMatch.index + urlMatch[0].length } });

  const emailMatch = text.match(emailPattern);
  if (emailMatch?.index !== undefined) signals.push({ kind: 'email', confidence: 'medium', range: { start: emailMatch.index, end: emailMatch.index + emailMatch[0].length } });

  const colorMatch = text.match(hexColorPattern);
  if (colorMatch?.index !== undefined) {
    const offset = colorMatch[1] ? colorMatch[1].length : 0;
    const color = colorMatch[0].slice(offset);
    signals.push({ kind: 'hex_color', confidence: 'medium', range: { start: colorMatch.index + offset, end: colorMatch.index + offset + color.length }, metadata: { value: color } });
  }

  if (markdownPattern.test(text)) signals.push({ kind: 'markdown', confidence: 'medium' });
  if (/```[\s\S]*?```/.test(text)) signals.push({ kind: 'code_block', confidence: 'medium' });
  if ((isCommandSnippet(text) || shellPattern.test(text)) && !signals.some((signal) => signal.kind === 'shell')) signals.push({ kind: 'shell', confidence: isCommandSnippet(text) ? 'high' : 'medium', language: 'shell' });

  return dedupeSignals(signals);
}

function isJson(value: string): boolean {
  try {
    JSON.parse(value);
    return true;
  } catch {
    return false;
  }
}

function normalizePlatform(platform: NodeJS.Platform | ClipboardPlatform | undefined): ClipboardPlatform {
  if (platform === 'darwin' || platform === 'win32' || platform === 'linux') return platform;
  return 'unknown';
}

function dedupeSignals(signals: ContentSignal[]): ContentSignal[] {
  const seen = new Set<string>();
  return signals.filter((signal) => {
    const key = `${signal.kind}:${signal.language ?? ''}:${signal.range?.start ?? ''}:${signal.range?.end ?? ''}`;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

function findJsonFragment(text: string): { start: number; end: number } | null {
  for (let start = 0; start < text.length; start += 1) {
    const opener = text[start];
    if (opener !== '{' && opener !== '[') continue;
    const closer = opener === '{' ? '}' : ']';
    let depth = 0;
    let inString = false;
    let escaped = false;
    for (let index = start; index < text.length; index += 1) {
      const char = text[index];
      if (inString) {
        if (escaped) {
          escaped = false;
        } else if (char === '\\') {
          escaped = true;
        } else if (char === '"') {
          inString = false;
        }
        continue;
      }
      if (char === '"') {
        inString = true;
      } else if (char === opener) {
        depth += 1;
      } else if (char === closer) {
        depth -= 1;
        if (depth === 0) {
          const end = index + 1;
          if (end - start < text.trim().length && isJson(text.slice(start, end))) return { start, end };
          break;
        }
      }
    }
  }
  return null;
}

File: src-tauri/src/detection.rs
use crate::models::{ClipboardFormatInfo, ContentSignal};

pub struct TypeInfo {
    pub item_type: String,
    pub code_language: Option<String>,
    pub url: Option<String>,
}

pub fn detect_clipboard_type(
    text: &str,
    html: Option<&str>,
    has_image: bool,
    file_paths: &[String],
) -> TypeInfo {
    let trimmed = detection_sample(text).trim();
    if has_image {
        return TypeInfo {
            item_type: "image".into(),
            code_language: None,
            url: None,
        };
    }
    if !file_paths.is_empty() {
        return TypeInfo {
            item_type: "file_reference".into(),
            code_language: None,
            url: None,
        };
    }
    if is_url(trimmed) {
        let url = if trimmed.starts_with("www.") {
            format!("https://{trimmed}")
        } else {
            trimmed.into()
        };
        return TypeInfo {
            item_type: "url".into(),
            code_language: None,
            url: Some(url),
        };
    }
    if is_command_snippet(trimmed) {
        return TypeInfo {
            item_type: "command".into(),
            code_language: Some("shell".into()),
            url: None,
        };
    }
    let language = detect_code_language(trimmed);
    if language != "unknown" {
        return TypeInfo {
            item_type: "code".into(),
            code_language: Some(language),
            url: None,
        };
    }
    if html.is_some_and(|value| !value.trim().is_empty()) {
        return TypeInfo {
            item_type: "rich_text".into(),
            code_language: None,
            url: None,
        };
    }
    TypeInfo {
        item_type: "plain_text".into(),
        code_language: None,
        url: None,
    }
}

pub fn normalize_formats(
    raw_formats: Vec<String>,
    has_text: bool,
    has_html: bool,
    has_rtf: bool,
    has_image: bool,
    has_files: bool,
) -> ClipboardFormatInfo {
    let mut normalized = Vec::<String>::new();
    if has_text {
        normalized.push("text/plain".into());
    }
    if has_html {
        normalized.push("text/html".into());
    }
    if has_rtf {
        normalized.push("text/rtf".into());
    }
    if has_image {
        normalized.push("image".into());
    }
    if has_files {
        normalized.push("files".into());
    }
    ClipboardFormatInfo {
        raw_formats,
        normalized_formats: normalized,
        has_text,
        has_html,
        has_rtf,
        has_image,
        has_files,
        platform: std::env::consts::OS.into(),
    }
}

pub fn detect_content_signals(
    text: &str,
    html: Option<&str>,
    code_language: Option<&str>,
) -> Vec<ContentSignal> {
    let mut signals = Vec::new();
    if html.is_some_and(|value| !value.trim().is_empty()) {
        signals.push(signal("html", "high", None));
    }
    if let Some(language) = code_language {
        if language != "unknown" {
            signals.push(signal(
                if language == "shell" { "shell" } else { "code" },
                "high",
                Some(language),
            ));
            if language == "json" {
                signals.push(signal("json", "high", Some("json")));
            }
        }
    }
    let trimmed = detection_sample(text).trim();
    if trimmed.contains("://") || trimmed.starts_with("www.") {
        signals.push(signal(
            "url",
            if is_url(trimmed) { "high" } else { "medium" },
            None,
        ));
    }
    if trimmed.contains('@') && trimmed.contains('.') {
        signals.push(signal("email", "medium", None));
    }
    if trimmed.contains("```") {
        signals.push(signal("code_block", "medium", None));
    }
    signals
}

fn signal(kind: &str, confidence: &str, language: Option<&str>) -> ContentSignal {
    ContentSignal {
        kind: kind.into(),
        confidence: confidence.into(),
        language: language.map(str::to_string),
        range: None,
        metadata: None,
    }
}

fn is_url(text: &str) -> bool {
    !text.contains(char::is_whitespace)
        && (starts_with_ignore_ascii_case(text, "http://")
            || starts_with_ignore_ascii_case(text, "https://")
            || starts_with_ignore_ascii_case(text, "www."))
}

fn starts_with_ignore_ascii_case(value: &str, prefix: &str) -> bool {
    value
        .get(..prefix.len())
        .is_some_and(|candidate| candidate.eq_ignore_ascii_case(prefix))
}

fn detection_sample(text: &str) -> &str {
    const MAX_DETECTION_BYTES: usize = 64 * 1024;
    if text.len() <= MAX_DETECTION_BYTES {
        return text;
    }
    let mut end = MAX_DETECTION_BYTES;
    while !text.is_char_boundary(end) {
        end -= 1;
    }
    &text[..end]
}

fn is_command_snippet(text: &str) -> bool {
    let commands = [
        "npm", "pnpm", "yarn", "git", "cd", "ls", "mkdir", "rm", "cp", "mv", "curl", "wget",
        "docker", "kubectl", "ssh", "brew", "python", "node",
    ];
    let lines: Vec<_> = text
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .collect();
    !lines.is_empty()
        && lines.len() <= 4
        && lines
            .iter()
            .all(|line| commands.iter().any(|command| line.starts_with(command)))
}

fn detect_code_language(text: &str) -> String {
    if text.is_empty() {
        return "unknown".into();
    }
    if (text.starts_with('{') || text.starts_with('['))
        && serde_json::from_str::<serde_json::Value>(text).is_ok()
    {
        return "json".into();
    }
    if text.contains("<!doctype html") || (text.contains('<') && text.contains("</")) {
        return "html".into();
    }
    if text.contains("function")
        || text.contains("const ")
        || text.contains("=>")
        || text.contains("console.log")
    {
        return "javascript".into();
    }
    if text.contains("interface ") || text.contains("type ") && text.contains("export ") {
        return "typescript".into();
    }
    if text.contains("def ") || text.contains("import ") && text.contains("print(") {
        return "python".into();
    }
    "unknown".into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classifies_large_content_from_a_bounded_sample() {
        let text = format!("const value = 1;{}", "x".repeat(2 * 1024 * 1024));
        let result = detect_clipboard_type(&text, None, false, &[]);
        assert_eq!(result.item_type, "code");
        assert_eq!(result.code_language.as_deref(), Some("javascript"));
    }

    #[test]
    fn keeps_url_detection_case_insensitive_without_allocating_a_lowercase_copy() {
        let result = detect_clipboard_type("HTTPS://example.com", None, false, &[]);
        assert_eq!(result.item_type, "url");
    }
}
